contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/DefaultInsertionCostCalculator.java

```diff
@@ -55,6 +60,32 @@ public double calculate(DrtRequest drtRequest, Insertion insertion, DetourTimeIn
 			return INFEASIBLE_SOLUTION_COST;
 		}
 
+		// all stops after the new (potential) pickup but before the new dropoff
+		// are delayed by pickupDetourTimeLoss
+		double detour = detourTimeInfo.pickupDetourInfo.pickupTimeLoss;
+		if(vEntry.stops != null) {
+			for (int s = insertion.pickup.index; s < vEntry.stops.size(); s++) {
+				Waypoint.Stop stop = vEntry.stops.get(s);
+				// passengers are being dropped off == may be close to arrival
+				if (!stop.task.getDropoffRequests().isEmpty()) {
+					double nextArrival = stop.getArrivalTime();
+					double departureTime = insertion.vehicleEntry.start.getDepartureTime();
+					//arrival is very soon
+					if (nextArrival - departureTime < drtConfigGroup.allowDetourBeforeArrivalThreshold &&
+							detour > 0) {
+						return INFEASIBLE_SOLUTION_COST;
+					} else {
+						// all following stops are above the threshold
```

Comment:
Not sure if this is correct if `detour == 0` and we are still within the time threshold.

If this is indeed incorrect, consider adding such a case to your tests (BTW. Nice that you have already added some tests!).